File: clients/android/app/src/main/kotlin/com/lextures/android/features/checklist/CourseChecklistSection.kt
package com.lextures.android.features.checklist

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.KeyboardArrowRight
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.MoreVert
import androidx.compose.material.icons.filled.Refresh
import androidx.compose.material.icons.filled.WifiOff
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.lextures.android.R
import com.lextures.android.core.auth.AuthSession
import com.lextures.android.core.design.accentColor
import com.lextures.android.core.design.textPrimary
import com.lextures.android.core.design.textSecondary
import com.lextures.android.core.i18n.L
import com.lextures.android.core.i18n.LocalLocalePreferences
import com.lextures.android.core.lms.ChecklistApi
import com.lextures.android.core.lms.ChecklistDismissReason
import com.lextures.android.core.lms.ChecklistItem
import com.lextures.android.core.lms.ChecklistNavTarget
import com.lextures.android.core.lms.CourseChecklist
import com.lextures.android.core.lms.CourseChecklistLogic
import com.lextures.android.core.lms.CourseChecklistSummaryStore
import com.lextures.android.core.lms.CourseChecklistTargetTable
import com.lextures.android.core.lms.CourseSummary
import com.lextures.android.core.network.ApiError
import com.lextures.android.core.routing.LinkOpener
import com.lextures.android.features.home.HomeShellState
import com.lextures.android.features.home.LmsEmptyState
import com.lextures.android.features.home.LmsErrorBanner
import kotlinx.coroutines.launch

@Composable
fun CourseChecklistSection(
    session: AuthSession,
    course: CourseSummary,
    shell: HomeShellState?,
    isOnline: Boolean,
    initialFocus: String? = null,
    modifier: Modifier = Modifier,
) {
    val context = LocalContext.current
    val localePrefs = LocalLocalePreferences.current
    val scope = rememberCoroutineScope()
    val table = remember { CourseChecklistTargetTable.shared(context) }

    var checklist by remember { mutableStateOf<CourseChecklist?>(null) }
    var loading by remember { mutableStateOf(false) }
    var errorMessage by remember { mutableStateOf<String?>(null) }
    var rateLimitMessage by remember { mutableStateOf<String?>(null) }
    var actionError by remember { mutableStateOf<String?>(null) }
    var showCompleted by remember { mutableStateOf(false) }
    var expandedCategories by remember { mutableStateOf(setOf<String>()) }
    var expandedItems by remember { mutableStateOf(setOf<String>()) }
    var dismissTarget by remember { mutableStateOf<ChecklistItem?>(null) }

    fun errorText(e: Exception): String = when (e) {
        is ApiError.HttpStatus -> e.apiMessage ?: L.text(context, localePrefs, R.string.mobile_checklist_loadError)
        else -> e.message ?: L.text(context, localePrefs, R.string.mobile_checklist_loadError)
    }

    suspend fun loadFull(force: Boolean = false) {
        if (!isOnline) return
        val token = session.accessToken.value ?: return
        if (!force && checklist != null) return
        loading = true
        errorMessage = null
        try {
            val result = ChecklistApi.fetchChecklist(course.courseCode, token)
            checklist = result
            CourseChecklistSummaryStore.applyChecklist(result)
            expandedCategories = result.categories
                .filter { CourseChecklistLogic.outstandingCount(it) > 0 }
                .map { it.id }
                .toSet()
        } catch (e: ApiError.HttpStatus) {
            if (e.code == 403) {
                CourseChecklistSummaryStore.markForbidden(course.courseCode)
                checklist = null
                errorMessage = null
            } else {
                errorMessage = errorText(e)
            }
        } catch (e: Exception) {
            errorMessage = errorText(e)
        } finally {
            loading = false
        }
    }

    LaunchedEffect(course.courseCode, isOnline) {
        loadFull()
    }

    fun openTarget(target: ChecklistNavTarget?) {
        val resolved = CourseChecklistLogic.resolveTarget(target, course.courseCode, table)
        when (resolved.kind) {
            CourseChecklistLogic.TargetKind.Native -> {
                resolved.workspaceSection?.let { shell?.activeCourseSection = it }
            }
            CourseChecklistLogic.TargetKind.Web -> {
                resolved.webPath?.let { path ->
                    val url = com.lextures.android.core.config.AppConfiguration.webUrl(path)
                    LinkOpener.open(context, url, shell, "checklist")
                }
            }
            CourseChecklistLogic.TargetKind.Unresolved -> {
                resolved.workspaceSection?.let { shell?.activeCourseSection = it }
                    ?: resolved.webPath?.let { path ->
                        val url = com.lextures.android.core.config.AppConfiguration.webUrl(path)
                        LinkOpener.open(context, url, shell, "checklist")
                    }
            }
        }
    }

    Column(modifier = modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(12.dp)) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Text(
                text = L.text(context, localePrefs, R.string.mobile_checklist_pageTitle),
                fontSize = 18.sp,
                fontWeight = FontWeight.SemiBold,
                color = textPrimary(),
                modifier = Modifier.weight(1f),
            )
            IconButton(
                onClick = {
                    scope.launch {
                        val token = session.accessToken.value ?: return@launch
                        if (!isOnline) return@launch
                        loading = true
                        rateLimitMessage = null
                        try {
                            val result = ChecklistApi.refreshChecklist(course.courseCode, token)
                            checklist = result
                            CourseChecklistSummaryStore.applyChecklist(result)
                        } catch (e: ApiError.HttpStatus) {
                            if (e.code == 429) {
                                rateLimitMessage = CourseChecklistLogic.RATE_LIMITED_MESSAGE
                            } else {
                                errorMessage = errorText(e)
                            }
                        } catch (e: Exception) {
                            errorMessage = errorText(e)
                        } finally {
                            loading = false
                        }
                    }
                },
                enabled = isOnline && !loading,
            ) {
                Icon(Icons.Default.Refresh, contentDescription = L.text(context, localePrefs, R.string.mobile_checklist_recheck))
            }
        }

        val summary = checklist?.summary ?: CourseChecklistSummaryStore.cached(course.courseCode)
        if (summary != null) {
            val progressLabel = CourseChecklistLogic.progressLabel(summary.done, summary.total)
            Text(progressLabel, color = textSecondary(), fontSize = 14.sp)
            LinearProgressIndicator(
                progress = { CourseChecklistLogic.progressFraction(summary.done, summary.total).toFloat() },
                modifier = Modifier
                    .fillMaxWidth()
                    .height(6.dp)
                    .clip(RoundedCornerShape(3.dp)),
                color = accentColor(),
            )
        }

        when {
            !isOnline -> {
                LmsEmptyState(
                    icon = Icons.Default.WifiOff,
                    title = L.text(context, localePrefs, R.string.mobile_checklist_offlineTitle),
                    message = L.text(context, localePrefs, R.string.mobile_checklist_offlineBody),
                )
            }
            errorMessage != null -> {
                Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                    LmsErrorBanner(errorMessage!!)
                    OutlinedButton(onClick = { scope.launch { loadFull(force = true) } }) {
                        Text(L.text(context, localePrefs, R.string.mobile_checklist_retry))
                    }
                }
            }
            loading && checklist == null -> {
                CircularProgressIndicator(modifier = Modifier.padding(24.dp))
            }
            checklist != null -> {
                val data = checklist!!
                val allDone = data.summary.outstandingTotal == 0 && data.summary.total > 0
                if (allDone && !showCompleted && data.dismissed.isEmpty()) {
                    Column(
                        modifier = Modifier
                            .fillMaxWidth()
                            .clip(RoundedCornerShape(12.dp))
                            .padding(12.dp),
                        verticalArrangement = Arrangement.spacedBy(8.dp),
                    ) {
                        Text(
                            L.text(context, localePrefs, R.string.mobile_checklist_allDoneTitle),
                            fontWeight = FontWeight.SemiBold,
                            color = textPrimary(),
                        )
                        Text(
                            L.text(context, localePrefs, R.string.mobile_checklist_allDoneBody),
                            color = textSecondary(),
                        )
                        OutlinedButton(onClick = { showCompleted = true }) {
                            Text(L.text(context, localePrefs, R.string.mobile_checklist_showCompleted))
                        }
                    }
                } else if (data.categories.isEmpty() && data.dismissed.isEmpty()) {
                    Text(
                        L.text(context, localePrefs, R.string.mobile_checklist_catalogEmpty),
                        color = textSecondary(),
                    )
                } else {
                    // Offer show/hide whenever any items are done (not only when everything is done).
                    if (data.summary.done > 0) {
                        TextButton(onClick = { showCompleted = !showCompleted }) {
                            Text(
                                L.text(
                                    context,
                                    localePrefs,
                                    if (showCompleted) R.string.mobile_checklist_hideCompleted
                                    else R.string.mobile_checklist_showCompleted,
                                ),
                            )
                        }
                    }
                    data.categories.forEach { category ->
                        val outstanding = CourseChecklistLogic.outstandingCount(category)
                        val expanded = category.id in expandedCategories || outstanding > 0
                        val items = CourseChecklistLogic.visibleItems(category, showCompleted)
                        if (items.isEmpty()) return@forEach
                        Column(verticalArrangement = Arrangement.spacedBy(6.dp)) {
                            Row(
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .clickable {
                                        expandedCategories = if (category.id in expandedCategories) {
                                            expandedCategories - category.id
                                        } else {
                                            expandedCategories + category.id
                                        }
                                    }
                                    .padding(vertical = 6.dp),
                                verticalAlignment = Alignment.CenterVertically,
                            ) {
                                Text(
                                    text = category.title,
                                    fontWeight = FontWeight.SemiBold,
                                    color = textPrimary(),
                                    modifier = Modifier.weight(1f),
                                )
                                if (outstanding > 0) {
                                    Text(
                                        text = context.getString(R.string.mobile_checklist_outstandingCount, outstanding),
                                        color = textSecondary(),
                                        fontSize = 12.sp,
                                    )
                                }
                            }
                            if (expanded) {
                                items.forEach { item ->
                                    ChecklistItemCard(
                                        item = item,
                                        expanded = item.id in expandedItems,
                                        isOnline = isOnline,
                                        onToggleEvidence = {
                                            expandedItems = if (item.id in expandedItems) {
                                                expandedItems - item.id
                                            } else {
                                                expandedItems + item.id
                                            }
                                        },
                                        onOpen = { openTarget(item.target) },
                                        onDismiss = { dismissTarget = item },
                                        onRecheck = {
                                            scope.launch {
                                                val token = session.accessToken.value ?: return@launch
                                                try {
                                                    ChecklistApi.recheckItem(course.courseCode, item.id, token)
                                                    loadFull(force = true)
                                                } catch (e: Exception) {
                                                    actionError = errorText(e)
                                                }
                                            }
                                        },
                                        onOpenEvidence = { openTarget(it) },
                                    )
                                }
                            }
                        }
                    }
                    if (data.dismissed.isNotEmpty()) {
                        Text(
                            text = context.getString(R.string.mobile_checklist_dismissedSection, data.dismissed.size),
                            fontWeight = FontWeight.SemiBold,
                            color = textPrimary(),
                            modifier = Modifier.padding(top = 8.dp),
                        )
                        data.dismissed.forEach { item ->
                            Row(
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .padding(vertical = 6.dp),
                                verticalAlignment = Alignment.CenterVertically,
                            ) {
                                Text(item.title, modifier = Modifier.weight(1f), color = textPrimary())
                                TextButton(
                                    onClick = {
                                        scope.launch {
                                            val token = session.accessToken.value ?: return@launch
                                            try {
                                                ChecklistApi.restoreItem(course.courseCode, item.id, token)
                                                loadFull(force = true)
                                            } catch (e: Exception) {
                                                actionError = errorText(e)
                                            }
                                        }
                                    },
                                    enabled = isOnline,
                                ) {
                                    Text(L.text(context, localePrefs, R.string.mobile_checklist_restore))
                                }
                            }
                        }
                    }
                }
            }
        }

        rateLimitMessage?.let {
            Text(it, color = textSecondary(), fontSize = 13.sp)
        }
        actionError?.let {
            Text(it, color = com.lextures.android.core.design.LexturesColors.Coral, fontSize = 13.sp)
        }
    }

    dismissTarget?.let { item ->
        ChecklistDismissDialog(
            item = item,
            isOnline = isOnline,
            onDismissRequest = { dismissTarget = null },
            onConfirm = { reason, note ->
                dismissTarget = null
                scope.launch {
                    val token = session.accessToken.value ?: return@launch
                    try {
                        ChecklistApi.dismissItem(course.courseCode, item.id, reason, note, token)
                        loadFull(force = true)
                    } catch (e: Exception) {
                        actionError = errorText(e)
                    }
                }
            },
        )
    }

    // silence unused
    @Suppress("UNUSED_VARIABLE")
    val focus = initialFocus
}

@Composable
private fun ChecklistItemCard(
    item: ChecklistItem,
    expanded: Boolean,
    isOnline: Boolean,
    onToggleEvidence: () -> Unit,
    onOpen: () -> Unit,
    onDismiss: () -> Unit,
    onRecheck: () -> Unit,
    onOpenEvidence: (ChecklistNavTarget?) -> Unit,
) {
    val context = LocalContext.current
    val localePrefs = LocalLocalePreferences.current
    val done = CourseChecklistLogic.isDone(item.status)
    val evidenceCount = item.evidence?.rows?.size ?: 0
    var menuOpen by remember { mutableStateOf(false) }

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(12.dp))
            .padding(10.dp)
            .semantics {
                contentDescription = "${item.title}, ${CourseChecklistLogic.accessibilityStatusValue(item.status)}"
            },
        verticalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        Row(verticalAlignment = Alignment.Top) {
            if (done) {
                Icon(Icons.Default.CheckCircle, contentDescription = null, tint = accentColor(), modifier = Modifier.size(20.dp))
            }
            Column(
                modifier = Modifier
                    .weight(1f)
                    .clickable(enabled = !done || evidenceCount > 0) {
                        if (evidenceCount > 0) onToggleEvidence() else if (!done) onOpen()
                    }
                    .padding(start = 8.dp),
            ) {
                Text(
                    text = item.title,
                    fontWeight = FontWeight.Medium,
                    color = textPrimary(),
                    textDecoration = if (done) TextDecoration.LineThrough else null,
                )
                item.detail?.takeIf { it.isNotBlank() }?.let {
                    Text(it, color = textSecondary(), fontSize = 13.sp)
                }
            }
            IconButton(onClick = { menuOpen = true }) {
                Icon(Icons.Default.MoreVert, contentDescription = L.text(context, localePrefs, R.string.mobile_checklist_overflowMenu))
            }
            DropdownMenu(expanded = menuOpen, onDismissRequest = { menuOpen = false }) {
                if (!done) {
                    DropdownMenuItem(
                        text = { Text(L.text(context, localePrefs, R.string.mobile_checklist_open)) },
                        onClick = { menuOpen = false; onOpen() },
                    )
                    DropdownMenuItem(
                        text = { Text(L.text(context, localePrefs, R.string.mobile_checklist_dismiss)) },
                        onClick = { menuOpen = false; onDismiss() },
                        enabled = isOnline,
                    )
                }
                DropdownMenuItem(
                    text = { Text(L.text(context, localePrefs, R.string.mobile_checklist_recheckItem)) },
                    onClick = { menuOpen = false; onRecheck() },
                    enabled = isOnline,
                )
            }
        }
        if (evidenceCount > 0) {
            TextButton(onClick = onToggleEvidence) {
                Text(
                    if (expanded) {
                        L.text(context, localePrefs, R.string.mobile_checklist_hideEvidence)
                    } else {
                        context.getString(R.string.mobile_checklist_showEvidence, evidenceCount)
                    },
                )
            }
        }
        if (expanded) {
            item.evidence?.rows?.forEach { row ->
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .clickable { onOpenEvidence(row.target) }
                        .padding(vertical = 8.dp, horizontal = 8.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Column(modifier = Modifier.weight(1f)) {
                        Text(row.label, color = textPrimary())
                        row.sublabel?.takeIf { it.isNotBlank() }?.let {
                            Text(it, color = textSecondary(), fontSize = 12.sp)
                        }
                    }
                    Icon(Icons.AutoMirrored.Filled.KeyboardArrowRight, contentDescription = null, tint = textSecondary())
                }
            }
        }
    }
}

File: clients/ios/Lextures/Features/Checklist/CourseChecklistSection.swift
import SwiftUI

/// Course checklist workspace section (CC.9).
struct CourseChecklistSection: View {
    @Environment(AuthSession.self) private var session
    @Environment(AppShellModel.self) private var shell
    @Environment(\.colorScheme) private var colorScheme
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    let course: CourseSummary
    var initialFocus: String?

    @State private var controller: CourseChecklistController
    @State private var showCompleted = false
    @State private var expandedItems: Set<String> = []
    @State private var dismissTarget: ChecklistItem?

    init(course: CourseSummary, initialFocus: String? = nil) {
        self.course = course
        self.initialFocus = initialFocus
        _controller = State(initialValue: CourseChecklistController(courseCode: course.courseCode))
    }

    private var isOnline: Bool { NetworkMonitor.shared.isOnline }
    private var summary: CourseChecklistSummary? {
        controller.checklist?.summary
            ?? CourseChecklistSummaryStore.shared.cached(courseCode: course.courseCode)
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 14) {
            header

            if !isOnline {
                offlinePanel
            } else if let errorMessage = controller.errorMessage {
                errorPanel(errorMessage)
            } else if controller.loading && controller.checklist == nil {
                LMSSkeletonList(count: 4)
            } else if let checklist = controller.checklist {
                content(checklist)
            } else {
                LMSSkeletonList(count: 3)
            }

            if let rateLimitMessage = controller.rateLimitMessage {
                Text(rateLimitMessage)
                    .font(.footnote)
                    .foregroundStyle(LexturesTheme.textSecondary(for: colorScheme))
            }
            if let actionError = controller.actionError {
                Text(actionError)
                    .font(.footnote)
                    .foregroundStyle(LexturesTheme.coral)
            }
        }
        .task {
            await controller.loadFull(
                accessToken: session.accessToken,
                isOnline: isOnline,
                initialFocus: initialFocus,
                reduceMotion: reduceMotion
            )
        }
        .sheet(item: $dismissTarget) { item in
            ChecklistDismissSheet(
                item: item,
                isOnline: isOnline,
                onCancel: { dismissTarget = nil },
                onConfirm: { reason, note in
                    dismissTarget = nil
                    Task {
                        await controller.dismiss(
                            item: item,
                            reason: reason,
                            note: note,
                            accessToken: session.accessToken,
                            isOnline: isOnline
                        )
                    }
                }
            )
            .presentationDetents([.medium, .large])
        }
    }

    private var header: some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack {
                Text(L.text("mobile.checklist.pageTitle"))
                    .font(LexturesTheme.displayFont(18))
                    .foregroundStyle(LexturesTheme.textPrimary(for: colorScheme))
                Spacer()
                Button {
                    Task {
                        await controller.refresh(
                            accessToken: session.accessToken,
                            isOnline: isOnline
                        )
                    }
                } label: {
                    Image(systemName: "arrow.clockwise")
                }
                .disabled(!isOnline || controller.loading)
                .accessibilityLabel(L.text("mobile.checklist.recheck"))
            }

            if let summary {
                let progress = CourseChecklistLogic.progressLabel(done: summary.done, total: summary.total)
                Text(progress)
                    .font(.subheadline)
                    .foregroundStyle(LexturesTheme.textSecondary(for: colorScheme))
                    .accessibilityLabel(progress)
                    .accessibilityValue(
                        "\(Int(CourseChecklistLogic.progressFraction(done: summary.done, total: summary.total) * 100)) percent"
                    )

                ProgressView(
                    value: CourseChecklistLogic.progressFraction(done: summary.done, total: summary.total)
                )
                .tint(LexturesTheme.accent(for: colorScheme))
            }
        }
    }

    private var offlinePanel: some View {
        LMSEmptyState(
            systemImage: "wifi.slash",
            title: L.text("mobile.checklist.offlineTitle"),
            message: L.text("mobile.checklist.offlineBody")
        )
    }

    private func errorPanel(_ message: String) -> some View {
        VStack(alignment: .leading, spacing: 8) {
            Text(message)
                .foregroundStyle(LexturesTheme.coral)
            Button(L.text("mobile.checklist.retry")) {
                Task {
                    await controller.loadFull(
                        accessToken: session.accessToken,
                        isOnline: isOnline,
                        force: true,
                        initialFocus: initialFocus,
                        reduceMotion: reduceMotion
                    )
                }
            }
            .buttonStyle(.bordered)
        }
    }

    @ViewBuilder
    private func content(_ checklist: CourseChecklist) -> some View {
        let allDone = checklist.summary.outstandingTotal == 0 && checklist.summary.total > 0
        let hasCompleted = checklist.summary.done > 0
        if allDone && !showCompleted && checklist.dismissed.isEmpty {
            allDonePanel
        } else if checklist.categories.isEmpty && checklist.dismissed.isEmpty {
            Text(L.text("mobile.checklist.catalogEmpty"))
                .foregroundStyle(LexturesTheme.textSecondary(for: colorScheme))
        } else {
            if hasCompleted {
                Toggle(isOn: $showCompleted) {
                    Text(showCompleted
                          ? L.text("mobile.checklist.hideCompleted")
                          : L.text("mobile.checklist.showCompleted"))
                }
                .toggleStyle(.button)
            }

            ForEach(checklist.categories) { category in
                categoryBlock(category)
            }

            if !checklist.dismissed.isEmpty {
                CourseChecklistDismissedBlock(
                    items: checklist.dismissed,
                    isOnline: isOnline,
                    onRestore: { item in
                        Task {
                            await controller.restore(
                                item: item,
                                accessToken: session.accessToken,
                                isOnline: isOnline
                            )
                        }
                    }
                )
            }
        }
    }

    private var allDonePanel: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text(L.text("mobile.checklist.allDoneTitle"))
                .font(.headline)
            Text(L.text("mobile.checklist.allDoneBody"))
                .foregroundStyle(LexturesTheme.textSecondary(for: colorScheme))
            Button(L.text("mobile.checklist.showCompleted")) {
                showCompleted = true
            }
            .buttonStyle(.bordered)
        }
        .padding(12)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(LexturesTheme.cardBackground(for: colorScheme))
        .clipShape(RoundedRectangle(cornerRadius: 12, style: .continuous))
    }

    @ViewBuilder
    private func categoryBlock(_ category: ChecklistCategory) -> some View {
        let outstanding = CourseChecklistLogic.outstandingCount(in: category)
        let expanded = controller.expandedCategories.contains(category.id) || outstanding > 0
        let items = CourseChecklistLogic.visibleItems(in: category, showCompleted: showCompleted)

        if items.isEmpty {
            EmptyView()
        } else {
            VStack(alignment: .leading, spacing: 6) {
                Button {
                    if controller.expandedCategories.contains(category.id) {
                        controller.expandedCategories.remove(category.id)
                    } else {
                        controller.expandedCategories.insert(category.id)
                    }
                } label: {
                    HStack {
                        Image(systemName: expanded ? "chevron.down" : "chevron.right")
                            .font(.caption.weight(.semibold))
                        Text(category.title)
                            .font(.subheadline.weight(.semibold))
                            .foregroundStyle(LexturesTheme.textPrimary(for: colorScheme))
                        Spacer()
                        if outstanding > 0 {
                            Text(L.format("mobile.checklist.outstandingCount", outstanding))
                                .font(.caption)
                                .foregroundStyle(LexturesTheme.textSecondary(for: colorScheme))
                        }
                    }
                }
                .buttonStyle(.plain)
                .accessibilityAddTraits(.isHeader)

                if expanded {
                    ForEach(items) { item in
                        CourseChecklistItemRow(
                            item: item,
                            isOnline: isOnline,
                            expanded: expandedItems.contains(item.id),
                            isHighlighted: controller.highlightAnchor != nil
                                && item.target?.anchor == controller.highlightAnchor,
                            onToggleEvidence: { toggleItem(item.id) },
                            onOpen: { openTarget(item.target) },
                            onDismiss: { dismissTarget = item },
                            onRecheck: {
                                Task {
                                    await controller.recheck(
                                        item: item,
                                        accessToken: session.accessToken,
                                        isOnline: isOnline
                                    )
                                }
                            },
                            onOpenEvidenceRow: { openTarget($0) }
                        )
                    }
                }
            }
            .padding(.vertical, 4)
        }
    }

    private func toggleItem(_ id: String) {
        if expandedItems.contains(id) {
            expandedItems.remove(id)
        } else {
            expandedItems.insert(id)
        }
    }

    private func openTarget(_ target: ChecklistNavTarget?) {
        let resolved = CourseChecklistLogic.resolveTarget(
            target,
            courseCode: course.courseCode,
            table: CourseChecklistTargetTable.shared
        )
        switch resolved.kind {
        case .native:
            if let section = resolved.workspaceSection {
                shell.activeCourseSection = section
                if let anchor = resolved.focusAnchor {
                    controller.applyHighlight(anchor: anchor, reduceMotion: reduceMotion)
                }
            }
        case .web:
            if let path = resolved.webPath {
                openWebPath(path)
            }
        case .unresolved:
            if let section = resolved.workspaceSection {
                shell.activeCourseSection = section
            } else if let path = resolved.webPath {
                openWebPath(path)
            }
        }
    }

    private func openWebPath(_ path: String) {
        let url = AppConfiguration.webURL(path: path)
        LinkOpener.open(
            LinkOpener.Request(urlString: url.absoluteString, source: "checklist"),
            shell: shell
        )
    }
}
